certstore_linux.go

```diff
@@ -1,14 +1,321 @@
+	"crypto"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"unicode/utf16"
+	"unsafe"
+)
+
+// Map go hash algorithm identifiers to NSS's SECOidTags
+var goToNssAlg = map[crypto.Hash]C.SECOidTag{
+	crypto.SHA1:   C.SEC_OID_SHA1,
+	crypto.SHA256: C.SEC_OID_SHA256,
+	crypto.SHA384: C.SEC_OID_SHA384,
+	crypto.SHA512: C.SEC_OID_SHA512,
+}
+
+// nssStore is a wrapper around a C.NSSInitContext
+type nssStore C.NSSInitContext
+
+// Identities implements the Store interface.
+func (store nssStore) Identities() ([]Identity, error) {
+	var (
+		identities = make([]Identity, 0)
+		certs      = C.PK11_ListCerts(C.PK11CertListUser, nil)
+		node       *C.CERTCertListNode
+	)
+	if certs == nil {
+		return nil, fmt.Errorf("error listing user certificates: %s", C.GoString(C.GetErrorString()))
+	}
+	defer C.CERT_DestroyCertList(certs)
+	for node = C.CertListHead(certs); C.CertListEnd(node, certs) == 0; node = C.CertListNext(node) {
+		identity := nssIdentity(*C.CERT_DupCertificate(node.cert))
+		identities = append(identities, &identity)
+	}
+	return identities, nil
+}
+
+// nssIdentity is a wrapper around a C.CERTCertificate.
+type nssIdentity C.CERTCertificate
+
+// Signer implements the Identity interface.
+func (i *nssIdentity) Signer() (crypto.Signer, error) {
+	return i, nil
+}
+
+// Certificate implements the Identity interface.
+func (i *nssIdentity) Certificate() (*x509.Certificate, error) {
+	var (
+		der   = i.derCert
+		bytes = C.GoBytes(unsafe.Pointer(der.data), C.int(der.len))
+	)
+	cert, err := x509.ParseCertificate(bytes)
+	if err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+// CertificateChain implements the Identity interface.
+func (i *nssIdentity) CertificateChain() ([]*x509.Certificate, error) {
+	var (
+		cert  = C.CERTCertificate(*i)
+		certs = C.CERT_GetCertChainFromCert(&cert, C.PR_Now(), C.certUsageAnyCA)
+	)
+	if certs == nil {
+		return nil, fmt.Errorf("error building certificate chain: %s", C.GoString(C.GetErrorString()))
+	}
+	defer C.CERT_DestroyCertList(certs)
+
+	var (
+		certificates = make([]*x509.Certificate, 0)
+		node         *C.CERTCertListNode
+	)
+	for node = C.CertListHead(certs); C.CertListEnd(node, certs) == 0; node = C.CertListNext(node) {
+		var (
+			identity  = nssIdentity(*node.cert)
+			cert, err = identity.Certificate()
+		)
+		if err != nil {
+			return nil, errors.New("unable to parse certificate when building chain")
+		}
+		certificates = append(certificates, cert)
+	}
+	return certificates, nil
 }
 
-// Implement this function, just to silence other compiler errors.
+// Delete implements the Identity interface.
+func (i *nssIdentity) Delete() error {
+	var (
+		cert      = C.CERTCertificate(*i)
+		secstatus = C.PK11_DeleteTokenCertAndKey(&cert, nil)
+	)
+	if secstatus != C.SECSuccess {
+		return errors.New("failed to delete certificate & key pair")
+	}
+	return nil
+}
+
+// Close implements the Identity interface.
+func (i *nssIdentity) Close() {
+	cert := C.CERTCertificate(*i)
```

Comment:
In a few places, you're doing `C.CERTCertificate(*i)`, and then doing something with `&cert`. Am I correct in thinking that this is copying the struct and then working on the copy? If so, it seems that something like this `CERT_DestroyCertificate` call would only affect the reference count on the copy. It seems like this could be problematic if we were to then accidentally try using the original struct again?

I'm wondering if, instead, it would be possible to convert the `*nssIdentity` directly to a `*C.CERTCertificate`. Does something like `cert := (*C.CERTCertificate)(i)` work?

Reply:
I think you're right and indeed that does work. Amendments coming right up..